aptos-move/block-executor/src/code_cache_global_manager.rs

```diff
@@ -0,0 +1,572 @@
+        } else {
+            // We are not in the done state, this is an error.
+            alert_or_println!(
+                "Unable to mark ready, state: {:?}, previous: {:?}, current: {:?}",
+                state,
+                previous,
+                current
+            );
+            false
+        }
+    }
+
+    /// When in [State::Ready], runs different checks on cached modules and environment:
+    ///   1. If the environment is not initialized, or is different from the one in storage, it is
+    ///      re-initialized, and module caches are flushed.
+    ///   2. If too many struct names have been cached in re-indexing map in runtime environment,
+    ///      struct type caches and module caches are flushed.
+    ///   3. If module cache size is too large (in bytes), it is flushed.
+    /// The final environment and module caches are returned.
+    pub fn check_ready_and_get_caches(
+        &self,
+        storage_environment: AptosEnvironment,
+        config: &BlockExecutorModuleCacheLocalConfig,
+    ) -> Result<(AptosEnvironment, Arc<GlobalModuleCache<K, DC, VC, E>>), VMStatus> {
+        let state = self.state.lock();
+        if !matches!(state.deref(), State::Ready(_)) {
+            let msg = format!(
+                "Expected ready state to check caches, got {:?}",
+                state.deref()
+            );
+            return Err(VMStatus::error(
+                UNKNOWN_INVARIANT_VIOLATION_ERROR,
+                Some(msg),
+            ));
+        }
+
+        let environment = self.get_or_initialize_environment(storage_environment);
+        let module_cache = self.module_cache.clone();
+
+        // Check 1: struct re-indexing map is not too large. If it is, we flush the cache. Also, we
+        // need to flush modules because they store indices into re-indexing map.
+        let runtime_environment = environment.runtime_environment();
+        let struct_name_index_map_size = runtime_environment
+            .struct_name_index_map_size()
+            .map_err(|err| err.finish(Location::Undefined).into_vm_status())?;
+        if struct_name_index_map_size > config.max_struct_name_index_map_num_entries {
+            module_cache.flush_unsync();
+            runtime_environment.flush_struct_name_and_info_caches();
+        }
+
+        // Check 2: If the module cache is too big, flush it.
+        if module_cache.size_in_bytes() > config.max_module_cache_size_in_bytes {
+            module_cache.flush_unsync();
+        }
+
+        Ok((environment, module_cache))
+    }
+
+    /// If state is [State::Ready], changes it to [State::Executing] with the same value, returning
+    /// true. Otherwise, returns false indicating that state transition failed, also raising an
+    /// alert.
+    pub fn mark_executing(&self) -> bool {
+        let mut state = self.state.lock();
+        if let State::Ready(v) = state.deref_mut() {
+            *state = State::Executing(mem::take(v));
+            true
```

Comment:
skis and snowboards, some people like early returns, some prefer this way with else :) I'm okay with this/either just pointing out.

Reply:
Had early return before, do not know why changed to if-else😄